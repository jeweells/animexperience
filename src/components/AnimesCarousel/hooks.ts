import { useLayoutEffect, useMemo, useState } from "react";
import useResizeObserver from "use-resize-observer";

export const useSliding = (
    gap: number,
    countElements: number
) => {
    const { ref: containerRef, width: containerWidth = 1 } = useResizeObserver<HTMLDivElement>();
    const { ref: scrollerRef, width: scrollerWidth = 1 } = useResizeObserver<HTMLDivElement>();
    const [totalInViewport, setTotalInViewport] = useState(0);
    const [viewed, setViewed] = useState(0);
    const elementWidth = useMemo(() => {
        return (containerWidth - gap * (countElements - 1)) / countElements;
    }, [countElements, gap, containerWidth]);
    useLayoutEffect(() => {
        const elementWidthWithGap = elementWidth + gap;

        let visibleElementsCount = Math.floor(scrollerWidth / elementWidthWithGap);
        if (elementWidthWithGap * visibleElementsCount + elementWidth <= scrollerWidth) {
            visibleElementsCount++;
        }
        setTotalInViewport(visibleElementsCount);
    }, [scrollerWidth, elementWidth]);

    const handlePrev = () => {
        setViewed(viewed => Math.min(countElements, Math.max(viewed - totalInViewport, 0)));
    };

    const handleNext = () => {
        setViewed(viewed => Math.min(countElements, Math.max(viewed + totalInViewport, 0)));
    };

    const distance = useMemo(() => {
        const elementWithGap = elementWidth + gap;
        return elementWithGap * viewed;
    }, [viewed, elementWidth, scrollerWidth]);

    const slideProps = useMemo(() => ({
        style: { transform: `translate3d(${-distance}px, 0, 0)` }
    }), [distance]);



    const hasPrev = distance > 0;
    const hasNext = (viewed + totalInViewport) < countElements;

    return {
        scrollerRef,
        handlePrev,
        handleNext,
        slideProps,
        containerRef,
        hasPrev,
        hasNext
    };
};
