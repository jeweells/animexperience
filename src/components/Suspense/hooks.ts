import React from 'react'
import { Optional, FStatus } from '../../types'

export class Suspended extends Error {}

type UnmanagedPromise = { promise: Promise<any>; resolve(): void; reject(): void }
const createPromise = (): UnmanagedPromise => {
    const r: Partial<UnmanagedPromise> = {}
    r.promise = new Promise((resolve, reject) => {
        r.resolve = resolve
        r.reject = reject
    })
    return r as any as UnmanagedPromise
}

export const useAssertStatus = (status: Optional<FStatus>) => {
    const [up, setUp] = React.useState<UnmanagedPromise | null>(null)
    React.useLayoutEffect(() => {
        if (up === null) {
            if (status === 'idle' || status === 'loading') {
                setUp(createPromise())
            }
        } else if (status !== 'idle' && status !== 'loading') {
            if (status === 'failed') {
                up.reject()
            } else {
                up.resolve()
            }
        }
    }, [status])
    if (status === 'idle' || status === 'loading') {
        throw up.promise
    }
}
