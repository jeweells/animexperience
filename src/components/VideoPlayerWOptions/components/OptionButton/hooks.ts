import { useCallback, useMemo } from 'react'
import { playerOptions } from '../../../../../redux/reducers/playerOptions'
import { useAppDispatch, useAppSelector } from '../../../../../redux/store'
import { Optional } from '../../../../types'

export const usePlayerOption = (name: Optional<string>) => {
    const opts = useAppSelector((d) => d.playerOptions.options)
    const dispatch = useAppDispatch()
    const option = useMemo(() => opts?.find((d) => d.name === name), [opts, name])
    const __po = useAppSelector((d) => d.playerOptions)

    console.debug('Displaying options:', __po)
    return {
        option,
        prefer: useCallback(
            (value: boolean) => {
                name && dispatch(playerOptions.prefer({ name, value }))
            },
            [name, dispatch],
        ),
        use: useCallback(() => {
            name && dispatch(playerOptions.use(name))
        }, [name, dispatch]),
    }
}
