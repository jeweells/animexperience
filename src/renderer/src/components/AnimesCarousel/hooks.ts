import { useMemo, useState, HTMLAttributes } from 'react'
import useResizeObserver from 'use-resize-observer'

export const useSliding = (gap: number, countElements: number, prevButtonWidth: number) => {
  const { ref: containerRef, width: containerWidth = 1 } = useResizeObserver<HTMLDivElement>()
  const { ref: scrollerRef, width: scrollerWidth = 1 } = useResizeObserver<HTMLDivElement>()
  const [viewed, setViewed] = useState(0)
  const availableScrollerWidth = scrollerWidth - prevButtonWidth

  const elementWidth = useMemo(() => {
    return (containerWidth - gap * (countElements - 1)) / countElements
  }, [countElements, gap, containerWidth])
  const totalInViewport = useMemo(() => {
    const elementWidthWithGap = elementWidth + gap

    let visibleElementsCount = Math.floor(availableScrollerWidth / elementWidthWithGap)
    if (elementWidthWithGap * visibleElementsCount + elementWidth <= availableScrollerWidth) {
      visibleElementsCount++
    }
    return visibleElementsCount
  }, [availableScrollerWidth, elementWidth])

  const [sliding, setSliding] = useState<boolean>(false)
  const handlePrev = () => {
    if (!sliding) {
      setViewed((viewed) => Math.min(countElements, Math.max(viewed - totalInViewport, 0)))
      setSliding(true)
    }
  }

  const handleNext = () => {
    if (!sliding) {
      setViewed((viewed) => Math.min(countElements, Math.max(viewed + totalInViewport, 0)))
      setSliding(true)
    }
  }

  const distance = useMemo(() => {
    const elementWithGap = elementWidth + gap
    return -prevButtonWidth + elementWithGap * viewed
  }, [viewed, elementWidth, gap, prevButtonWidth])

  const slideProps = useMemo<Partial<HTMLAttributes<HTMLElement>>>(
    () => ({
      style: {
        transform: `translate3d(${-distance}px, 0, 0)`,
        // These will be unset when the transition ends
        pointerEvents: sliding ? 'none' : 'all'
      }
    }),
    // Do not depend it on distance; very important! Otherwise, 1 frame will be inconsistent
    [sliding]
  )

  const hasPrev = distance > 0
  const hasNext = viewed + totalInViewport < countElements

  return {
    scrollerRef,
    handlePrev,
    handleNext,
    slideProps,
    containerRef,
    hasPrev,
    hasNext,
    sliding,
    viewed,
    totalInViewport,
    onSlidingStart: () => {
      setSliding(true)
    },
    onSlidingComplete: () => {
      setSliding(false)
    }
  }
}

const defaultSizes = {
  gap: 16,
  containerWidth: '100vw',
  navigationWidth: 72
}

export const useSizes = () => {
  return defaultSizes
}
