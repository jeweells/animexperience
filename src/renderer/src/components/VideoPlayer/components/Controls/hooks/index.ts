export * from './useControls'
export * from './useBuffer'
export * from './useLoading'
export * from './usePlayPause'
export * from './useSeek'
export * from './useShowControls'
export * from './useVolume'
