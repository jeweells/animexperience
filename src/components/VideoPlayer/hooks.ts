import React from "react";
import $ from "jquery";
import { RecentAnimeData } from "../../hooks/useRecentAnimes";
import { Optional } from "../../types";
import { VideoOption } from "./index";
type $IframeContents = JQuery<HTMLIFrameElement | Text | Comment | Document>;

// Target must be iframe.contents() ... always
function * deepIframes (
    target: $IframeContents,
    depth = 5,
    currDepth = 1
): Generator<$IframeContents> {
    if (!target || target.length === 0 || currDepth >= depth) return;
    if (currDepth === 1) {
        yield target;
    }
    const iframes = target.find("iframe").get().map(x => $(x).contents());
    for (const iframe of iframes) {
        yield iframe;
    }
    for (const iframe of iframes) {
        for (const t of deepIframes(iframe, depth, currDepth + 1)) {
            yield t;
        }
    }
}

// The array returned will always be of length greater than 0
const deepFindVideos = (target: $IframeContents, depth = 5): Optional<HTMLVideoElement[]> => {
    for (const iframe of deepIframes(target, depth)) {
        const video = iframe.find("video");
        if (video.length > 0) {
            return video.get();
        }
    }
};


const handleSpecificOptions = (option: Optional<VideoOption>, contents: $IframeContents): boolean => {
    const methods: Partial<Record<string, (contents: $IframeContents) => boolean>> = {
        "fembed": handleFembedPlayer,
        "okru": handleOkRuPlayer,
        "mixdrop": handleMixDrop,
    };
    const handler = methods[option?.name?.toLowerCase() ?? ""];
    if (handler) {
        for (const iframe of deepIframes(contents)) {
            if (handler(iframe)) {
                return true;
            }
        }
    }

    return false;
};

const handleFembedPlayer = (iframe: $IframeContents) => {
    console.debug("Fembed: Clicking play button");
    const playBtn = iframe.find(".faplbu");
    if (playBtn.length > 0) {
        console.debug("GOT", playBtn);
        playBtn.trigger("click");
        return true;
    }
    return false;
};

const handleOkRuPlayer = (iframe: $IframeContents) => {
    console.debug("Okru: Clicking play button");
    const playBtn = iframe.find("div#embedVideoC.vid-card_cnt_w");
    if (playBtn.length > 0) {
        console.debug("GOT", playBtn);
        playBtn.trigger("click");
        // Let it spam or else it won't work
        return false;
    }
    return false;
};

const handleMixDrop = (iframe: $IframeContents) => {
    let adClicked = false;
    for (const adClick of iframe.find("div[onclick]").get()) {
        const onclick = adClick.getAttribute("onclick");
        if (onclick && onclick.includes("$(this).remove()")) {
            $(adClick).trigger("click");
            adClicked = true;
        }
    }
    if (!adClicked) {
        iframe.find("button.vjs-big-play-button").trigger("click");
    }
    return false;
};

export type BasicVideoInfo = {
    option?: Optional<VideoOption>;
    anime?: Optional<RecentAnimeData>;
}

export const useVideo = (info: BasicVideoInfo, container: Optional<HTMLDivElement>, ms = 300) => {
    const [video, setVideo] = React.useState<Optional<HTMLVideoElement>>(null);
    React.useLayoutEffect(() => {
        if (container) {
            const interval: { handle?: number; handledSpecificOptions?: boolean } = {};
            const iframeContent = $(container).find("iframe");
            const check = () => {
                const contents = iframeContent.contents();
                if (!interval.handledSpecificOptions) {
                    interval.handledSpecificOptions = handleSpecificOptions(info.option, contents);
                }
                const video = deepFindVideos(contents);
                console.debug("Trying to find video: ", video);
                const targetVideo = video?.find(vid => {
                    const jVid = $(vid);
                    if (jVid.attr("src")) {
                        return true;
                    } else if (jVid.find("source[src]").length) {
                        return true;
                    }
                    return false;
                });
                if (targetVideo) {
                    setVideo(targetVideo);
                    clearInterval(interval.handle);
                }
            };
            interval.handle = setInterval(check, ms);
            check();
            return () => {
                clearInterval(interval.handle);
            };
        }
    }, [info.anime?.name, info.anime?.episode, info.option?.name, container, ms]);

    return video;
};


export const useVideoImprovements = (info: BasicVideoInfo, container: Optional<HTMLDivElement>) => {
    const video = useVideo(info, container);
    React.useLayoutEffect(() => {
        if (video) {
            $(video).attr("autoplay", "true");
            video.play();
            const handleTimeUpdate = () => {
                console.debug("VIDEO TIME", video.currentTime);
            };
            video.addEventListener("timeupdate", handleTimeUpdate);
            return () => {
                video.removeEventListener("timeupdate", handleTimeUpdate);
            };
        }
    }, [video]);
};
