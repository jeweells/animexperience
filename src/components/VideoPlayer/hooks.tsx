import $ from 'jquery'
import React from 'react'
import ReactDOM from 'react-dom'
import { Provider } from 'react-redux'
import { EpisodeInfo, Store } from '../../../globals/types'
import { followedAnimes } from '../../../redux/reducers/followedAnimes'
import { watch } from '../../../redux/reducers/watch'
import { watched } from '../../../redux/reducers/watched'
import store, { useAppDispatch, useAppSelector } from '../../../redux/store'
import { useStaticStore } from '../../hooks/useStaticStore'
import { useWatched } from '../../hooks/useWatched'
import { Optional } from '../../types'
import NextEpisodeButton from '../NextEpisodeButton'
import { handleSpecificOptions } from './autoPlayHandlers'
import {
    RATIO_TO_FOLLOW_AN_ANIME,
    SECONDS_LEFT_TO_NEXT_EPISODE,
    SECONDS_LEFT_TO_TRIGGER_NEXT_EPISODE,
    SECONDS_NEXT_BUTTON_DISPLAY,
} from './constants'
import { handleFullScreen } from './fullScreenHandlers'
import { handleSeek } from './seekHandlers'
import { BasicVideoInfo } from './types'
import { deepFindVideos } from './utils'

export const useVideo = (
    info: BasicVideoInfo,
    container: Optional<HTMLDivElement>,
    ms = 300,
) => {
    const [video, setVideo] = React.useState<Optional<HTMLVideoElement>>(null)
    const [detachedVideo, setDetachedVideo] =
        React.useState<Optional<HTMLVideoElement>>(null)
    const watchEpisodeStatus = useAppSelector((d) => d.watch.status.watchEpisode)
    const dispatch = useAppDispatch()
    const episodeInfo = useWatched(info.anime)

    React.useLayoutEffect(() => {
        if (watchEpisodeStatus === 'succeeded' && container && episodeInfo !== null) {
            const refs: {
                handle?: number
                fsHandle?: number
                handledSpecificOptions?: boolean
                handledFullscreen?: boolean
            } = {}
            const iframeContent = $(container).find('iframe')
            const check = () => {
                const contents = iframeContent.contents()
                if (!refs.handledSpecificOptions) {
                    refs.handledSpecificOptions = handleSpecificOptions(
                        info.option,
                        contents,
                        episodeInfo,
                    )
                }
                const video = deepFindVideos(contents)
                const targetVideo = video?.find((vid) => {
                    const jVid = $(vid)
                    if (jVid.attr('src')) {
                        return true
                    } else if (jVid.find('source[src]').length) {
                        return true
                    }
                    return false
                })
                if (targetVideo) {
                    setVideo(targetVideo)
                    const jTargetVideo = $(targetVideo)
                    jTargetVideo.on('DOMNodeRemoved', () => {
                        if (jTargetVideo.parent().length === 0) {
                            console.debug('VIDEO DETACHED')
                            setVideo((p) => {
                                if (p === targetVideo) {
                                    jTargetVideo.off('DOMNodeRemoved', '**')
                                    return null
                                }
                                return p
                            })
                            setDetachedVideo(targetVideo)
                        }
                    })
                    clearInterval(refs.handle)
                }
            }
            const checkFs = () => {
                const contents = iframeContent.contents()
                if (!refs.handledFullscreen) {
                    refs.handledFullscreen = handleFullScreen(
                        info.option,
                        contents,
                        episodeInfo,
                    )
                }
                if (document.fullscreenElement) {
                    clearInterval(refs.fsHandle)
                }
            }
            refs.handle = setInterval(check, ms)
            check()
            const autoFullScreen = store.getState().watch.autoFullScreen
            console.debug('Displaying with fullscreen:', autoFullScreen)
            if (autoFullScreen) {
                // Disable auto fullscreen until the next episode is triggered
                dispatch(watch.setAutoFullScreen(false))
                refs.fsHandle = setInterval(checkFs, ms)
                checkFs()
            }
            return () => {
                clearInterval(refs.handle)
                clearInterval(refs.fsHandle)
            }
        }
    }, [
        info.anime?.name,
        info.anime?.episode,
        info.option?.name,
        container,
        ms,
        detachedVideo,
        episodeInfo === null,
        watchEpisodeStatus,
    ])
    return video
}

export const useVideoImprovements = (
    info: BasicVideoInfo,
    container: Optional<HTMLDivElement>,
) => {
    const video = useVideo(info, container)
    const dispatch = useAppDispatch()
    const staticStore = useStaticStore(Store.WATCHED)
    React.useLayoutEffect(() => {
        if (video) {
            if (!video.autoplay) {
                video.autoplay = true
            }
            const anime = info.anime
            if (anime) {
                staticStore
                    .get(anime.name, anime.episode)
                    .then((data: Optional<EpisodeInfo>) => {
                        // Try auto play the video at latest time stored
                        handleSeek(info, data, video)
                    })
            }
            const refs: {
                nextButtonShown?: boolean
                nextBtnRef?: JQuery | null
                watched?: boolean
                followed?: boolean
            } = {}
            const handleTimeUpdate = (e: Event) => {
                console.debug('TARGET', e.target, e.currentTarget)
                if (isFinite(video.duration)) {
                    const { duration, currentTime } = video
                    const currentTimeRatio = duration === 0 ? 0 : currentTime / duration

                    console.debug('VIDEO TIME', video.currentTime)
                    if (currentTime + SECONDS_LEFT_TO_NEXT_EPISODE >= duration) {
                        if (!refs.watched) {
                            refs.watched = true
                            // At this time of the video, the video can be considered "watched"
                            dispatch(watched.updateRecentlyWatched(anime))
                        }
                        const timeoutToNextEpisode = Math.max(
                            0,
                            duration -
                                (currentTime + SECONDS_LEFT_TO_TRIGGER_NEXT_EPISODE),
                        )
                        if (timeoutToNextEpisode === 0) {
                            // With timeout -1 auto play next episode is disabled (user probable moved the mouse)
                            const _state = store.getState()
                            if (_state.watch.nextEpisodeTimeout !== -1) {
                                dispatch(watch.nextEpisode())
                                refs.nextButtonShown = false
                                dispatch(watch.setNextEpisodeButton(false))
                            }
                        } else if (!refs.nextButtonShown) {
                            refs.nextButtonShown = true
                            dispatch(watch.setNextEpisodeButton(true))
                            dispatch(
                                watch.setNextEpisodeTimeout(SECONDS_NEXT_BUTTON_DISPLAY),
                            )
                        }
                    } else if (refs.nextButtonShown) {
                        refs.nextButtonShown = false
                        dispatch(watch.setNextEpisodeButton(false))
                    }
                    if (
                        anime &&
                        anime.name &&
                        typeof anime.episode === 'number' &&
                        // Save each 3 seconds
                        Math.floor(video.currentTime) % 3 === 0 &&
                        video.currentTime >= 3
                    ) {
                        dispatch(
                            watched.updateWatched({
                                anime,
                                info: {
                                    duration: video.duration,
                                    currentTime: video.currentTime,
                                    at: new Date().getTime(),
                                },
                            }),
                        )
                    }

                    if (currentTimeRatio >= RATIO_TO_FOLLOW_AN_ANIME && !refs.followed) {
                        console.debug('Able to follow this anime')
                        const watchState = store.getState().watch
                        if (watchState.info && watchState.watching) {
                            refs.followed = true
                            console.debug('Marking this anime as followed')
                            dispatch(
                                followedAnimes.follow({
                                    anime: watchState.watching,
                                    info: watchState.info,
                                }),
                            )
                        }
                    }
                }
            }
            console.debug('VIDEO', video)
            const handleMouseMove = () => {
                dispatch(watch.setNextEpisodeTimeout(-1))
            }
            const handleFullScreen = () => {
                // document.fullscreenElement will point to the element that
                // is in fullscreen mode if there is one. If there isn't one,
                // the value of the property is null.
                let targetElement: Optional<Element> = document.fullscreenElement
                // The target fullscreen element changes depending on the iframe we're in
                // It's needed to find an non iframe element to attach the button
                // so that its visible in fullscreen mode
                while (
                    targetElement?.tagName.toLowerCase() === 'iframe' &&
                    'contentDocument' in targetElement
                ) {
                    // targetElement instanceof HTMLIFrameElement does not work for iframes
                    // Alternative: Use iframe.contentWindow.HTMLIFrameElement
                    // However, this might be a cleaner way
                    targetElement = (targetElement as HTMLIFrameElement).contentDocument
                        ?.fullscreenElement
                }
                if (targetElement) {
                    const jTargetElement = $(targetElement as HTMLElement)
                    console.debug('Setting fullscreen')
                    const root = refs.nextBtnRef || $("<div id='raex-injected-root' />")
                    if (!root.parent().is(jTargetElement)) {
                        refs.nextBtnRef = root
                        console.debug('Attaching node to element')
                        jTargetElement.append(root)
                        ReactDOM.render(
                            <Provider store={store}>
                                <NextEpisodeButton />
                            </Provider>,
                            root.get(0),
                        )
                    } else {
                        console.debug('Element already attached')
                    }
                } else {
                    console.debug('Deleting node')
                    refs.nextBtnRef?.remove()
                    refs.nextBtnRef = null
                }
            }
            document.addEventListener('fullscreenchange', handleFullScreen)
            video.addEventListener('mousemove', handleMouseMove)
            video.addEventListener('timeupdate', handleTimeUpdate)
            return () => {
                video.removeEventListener('timeupdate', handleTimeUpdate)
                video.removeEventListener('mousemove', handleMouseMove)
                document.removeEventListener('fullscreenchange', handleFullScreen)
            }
        }
    }, [video])
}
