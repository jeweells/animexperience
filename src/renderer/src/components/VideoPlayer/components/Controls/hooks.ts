import { useContext, useEffect, useLayoutEffect, useState } from 'react'
import { Context } from './context'
import { useIsFullscreen } from '~/src/hooks/useIsFullscreen'
import { range } from '~/src/utils'

export const useControls = () => useContext(Context)

const MS_TO_HIDE = 4000

export const useShowControls = () => {
  const [show, setShow] = useState(true)
  const { isPlaying } = usePlayPause()
  const { loading } = useLoading()
  const { video } = useControls()

  useLayoutEffect(() => {
    if (!isPlaying) setShow(true)
    if (loading) setShow(true)
  }, [isPlaying, loading])

  useEffect(() => {
    if (!video) return
    const showPlayBar = () => {
      setShow(true)
    }
    document.addEventListener('mousemove', showPlayBar)
    video.addEventListener('mousemove', showPlayBar)
    return () => {
      video.removeEventListener('mousemove', showPlayBar)
      document.removeEventListener('mousemove', showPlayBar)
    }
  }, [video])

  useEffect(() => {
    if (!isPlaying) return
    if (loading) return
    if (!show) return
    const t = setTimeout(() => {
      setShow(false)
    }, MS_TO_HIDE)
    return () => clearTimeout(t)
  }, [show, isPlaying, loading])
  return { show, transitionMs: 500 }
}

export const usePlayPause = () => {
  const { video } = useControls()
  const [playing, setPlaying] = useState(false)

  useLayoutEffect(() => {
    if (video) return
    setPlaying(false)
  }, [video])

  useEffect(() => {
    if (!video) return
    const handlePlay = () => {
      setPlaying(true)
    }
    const handlePause = () => {
      setPlaying(false)
    }
    setPlaying(!video.paused)
    video.addEventListener('play', handlePlay)
    video.addEventListener('pause', handlePause)
    return () => {
      video.removeEventListener('play', handlePlay)
      video.removeEventListener('pause', handlePause)
    }
  }, [video])

  return {
    isPlaying: playing,
    play: () => video?.play() ?? Promise.resolve(),
    pause: () => video?.pause()
  }
}

export const useVolume = () => {
  const { video } = useControls()
  const [volume, setVolume] = useState(0)
  useLayoutEffect(() => {
    if (video) return
    setVolume(0)
  }, [video])

  useEffect(() => {
    if (!video) return
    const updateVolume = () => {
      setVolume(video.volume)
    }
    updateVolume()
    video.addEventListener('volumechange', updateVolume)
    return () => {
      video.removeEventListener('volumechange', updateVolume)
    }
  }, [video])
  return {
    volume,
    setVolume: (value: number) => {
      if (!video) return
      video.volume = value
      setVolume(value)
    }
  }
}

const defaultTime = { currentTime: 0, duration: 0 }

export const useSeek = () => {
  const { video } = useControls()
  const [time, setTime] = useState(defaultTime)

  useLayoutEffect(() => {
    if (video) return
    setTime(defaultTime)
  }, [video])

  useEffect(() => {
    if (!video) return
    const updateTime = () => {
      setTime({
        currentTime: video.currentTime,
        duration: video.duration
      })
    }
    updateTime()
    video.addEventListener('timeupdate', updateTime)
    return () => {
      video.removeEventListener('timeupdate', updateTime)
    }
  }, [video])
  return {
    time: video ? time : { currentTime: 0, duration: 0 },
    seek: (value: number) => {
      if (!video) return
      video.currentTime = value
    },
    fastSeek: (value: number) => {
      if (!video) return
      if (!video.fastSeek) video.currentTime = value
      else video.fastSeek(value)
    }
  }
}

export const useBuffer = () => {
  const { video } = useControls()

  type Buffer = {
    start: number
    end: number
  }

  const [buffered, setBuffered] = useState<Buffer[]>([])

  useLayoutEffect(() => {
    if (video) return
    if (buffered.length) setBuffered([])
  }, [video])

  useEffect(() => {
    if (!video) return
    const updateBuffer = () => {
      setBuffered(
        range(video.buffered.length).map((index) => {
          return {
            start: video.buffered.start(index) / video.duration,
            end: video.buffered.end(index) / video.duration
          }
        })
      )
    }
    updateBuffer()
    video.addEventListener('progress', updateBuffer)
    return () => {
      video.removeEventListener('progress', updateBuffer)
    }
  }, [video])
  return {
    buffered
  }
}

export const useLoading = () => {
  const { video } = useControls()

  const [loading, setLoading] = useState(true)

  useLayoutEffect(() => {
    if (video) return
    setLoading(true)
  }, [video])

  useEffect(() => {
    if (!video) return
    const seeking = () => {
      setLoading(video.seeking)
    }
    const seeked = () => {
      setLoading(video.seeking)
    }
    seeking()
    video.addEventListener('seeking', seeking)
    video.addEventListener('seeked', seeked)
    return () => {
      video.removeEventListener('seeking', seeking)
      video.removeEventListener('seeked', seeked)
    }
  }, [video])
  return {
    loading
  }
}

export const useFullscreen = () => {
  const isFullscreen = useIsFullscreen()

  return {
    isFullscreen,
    toggleFullscreen: () => {
      return window.setFullscreen(!isFullscreen)
    }
  }
}
