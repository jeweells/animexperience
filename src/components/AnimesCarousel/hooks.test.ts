import { useSliding } from './hooks'
import { renderHook, act } from '@testing-library/react-hooks'
import useResizeObserver from 'use-resize-observer'

describe('AnimesCarousel > useSliding', () => {
    const useResizeObserverMock = useResizeObserver as jest.Mock

    const snapshot = ({
        name,
        hook,
    }: {
        name?: string
        hook: ReturnType<typeof useSliding>
    }) => {
        const {
            scrollerRef,
            containerRef,
            onSlidingComplete,
            handleNext,
            handlePrev,
            ...props
        } = hook
        if (name) expect(props).toMatchSnapshot(name)
        else expect(props).toMatchSnapshot()
    }

    const getHook = (gap = 20, countElements = 4, prevButtonWidth = 50) => {
        return () => useSliding(gap, countElements, prevButtonWidth)
    }

    beforeEach(() => {
        useResizeObserverMock.mockReturnValue({ ref: jest.fn(), width: 100, height: 100 })
    })

    it('renders default', () => {
        const { result } = renderHook(getHook())
        snapshot({ hook: result.current })
    })

    it('renders when clicking next/prev', () => {
        const { result } = renderHook(getHook())

        act(() => {
            result.current.handleNext()
        })

        snapshot({ name: 'Clicked next', hook: result.current })

        act(() => {
            result.current.onSlidingComplete()
        })

        snapshot({ name: 'Finished sliding next', hook: result.current })

        act(() => {
            result.current.handlePrev()
        })

        snapshot({ name: 'Clicked prev', hook: result.current })

        act(() => {
            result.current.onSlidingComplete()
        })

        snapshot({ name: 'Finished sliding prev', hook: result.current })
    })
})
