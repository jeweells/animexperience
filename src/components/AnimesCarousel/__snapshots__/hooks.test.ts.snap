// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimesCarousel > useSliding renders default 1`] = `
Object {
  "hasNext": true,
  "hasPrev": false,
  "slideProps": Object {
    "style": Object {
      "pointerEvents": "all",
      "transform": "translate3d(50px, 0, 0)",
    },
  },
  "sliding": false,
  "totalInViewport": 2,
  "viewed": 0,
}
`;

exports[`AnimesCarousel > useSliding renders when clicking next/prev: Clicked next 1`] = `
Object {
  "hasNext": false,
  "hasPrev": true,
  "slideProps": Object {
    "style": Object {
      "pointerEvents": "none",
      "transform": "translate3d(-10px, 0, 0)",
    },
  },
  "sliding": true,
  "totalInViewport": 2,
  "viewed": 2,
}
`;

exports[`AnimesCarousel > useSliding renders when clicking next/prev: Clicked prev 1`] = `
Object {
  "hasNext": true,
  "hasPrev": false,
  "slideProps": Object {
    "style": Object {
      "pointerEvents": "none",
      "transform": "translate3d(50px, 0, 0)",
    },
  },
  "sliding": true,
  "totalInViewport": 2,
  "viewed": 0,
}
`;

exports[`AnimesCarousel > useSliding renders when clicking next/prev: Finished sliding next 1`] = `
Object {
  "hasNext": false,
  "hasPrev": true,
  "slideProps": Object {
    "style": Object {
      "pointerEvents": "all",
      "transform": "translate3d(-10px, 0, 0)",
    },
  },
  "sliding": false,
  "totalInViewport": 2,
  "viewed": 2,
}
`;

exports[`AnimesCarousel > useSliding renders when clicking next/prev: Finished sliding prev 1`] = `
Object {
  "hasNext": true,
  "hasPrev": false,
  "slideProps": Object {
    "style": Object {
      "pointerEvents": "all",
      "transform": "translate3d(50px, 0, 0)",
    },
  },
  "sliding": false,
  "totalInViewport": 2,
  "viewed": 0,
}
`;
